docs/source/best_practices.rst

```diff
@@ -0,0 +1,187 @@
+
+Say you wanted to rename all function calls which were calling global functions.
+So, you write the following::
+
+    class FunctionRenamer(cst.CSTTransformer):
+        def leave_Call(self, original_node: cst.Call, updated_node: cst.Call) -> cst.Call:
+            if m.matches(original_node.func, m.Name()):
+                return original_node.with_changes(
+                    func=cst.Name(
+                        "renamed_" + cst.ensure_type(original_node.func, cst.Name).value
+                    )
+                )
+            return original_node
+
+Consider writing instead::
+
+    class FunctionRenamer(cst.CSTTransformer):
+        def leave_Call(self, original_node: cst.Call, updated_node: cst.Call) -> cst.Call:
+            if m.matches(updated_node.func, m.Name()):
+                return updated_node.with_changes(
+                    func=cst.Name(
+                        "renamed_" + cst.ensure_type(updated_node.func, cst.Name).value
+                    )
+                )
+            return updated_node
+
+The version that returns modifications to ``original_node`` has a subtle bug. Consider
+the following code snippet::
+
+    some_func(1, 2, other_func(3))
+
+Running the recommended transform will return us a new code snippet that looks like this::
+
+    renamed_some_func(1, 2, renamed_other_func(3))
+
+However, running the version which modifies ``original_node`` will instead return::
+
+    renamed_some_func(1, 2, other_func(3))
+
+That's because the ``updated_node`` tree contains the modification to ``other_func``.
+By returning modifications to ``original_node`` instead of ``updated_node``, we accidentally
+discarded all the work done deeper in the tree.
+
+Provide a ``config`` when generating code from templates
+--------------------------------------------------------
+
+When generating complex trees it is often far easier to pass a f-string to
+:func:`~libcst.parse_statement` or :func:`~libcst.parse_expression` than it is to
+manually construct the tree. When using these functions to generate code, you should
+always use the ``config`` parameter in order to generate code that matches the
+defaults of the module you are modifying. The :class:`~libcst.Module` class even has
+a helper attribute :attr:`~libcst.Module.config_for_parsing` to make it easy to use. This
+ensures that line endings and indentation are consistent with the defaults in the
+module you are adding the code to.
+
+For example, to add a print statement to the end of a module::
+
+    module = cst.parse_module(some_code_string)
+    new_module = module.with_changes(
+        body=(
+            *module.body,
+            cst.parse_statement(
+                "print('Hello, world!')",
+                config=module.config_for_parsing,
+            ),
+        ),
+    )
+    new_code_string = new_module.code
+
+Leaving out the ``config`` parameter means that LibCST will assume some defaults
```

Comment:
Is it possible to have a simple example to demonstrate the caveat when user don't pass config?

Reply:
Not really. Any simpler examples don't have the caveat of parsing subtly incorrectly.